sb_vision/distance_finding.py

```diff
@@ -0,0 +1,121 @@
+"""
+Handlers for detecting 3D marker Location.
+
+Also handles loading the calibration file which is passed into the
+location finding function.
+"""
+
+import functools
+import re
+from pathlib import Path
+from typing import Any, Dict, List, NamedTuple, Tuple
+
+import cv2
+import numpy as np
+from lxml import etree
+
+Coordinate = NamedTuple('PixelCoordinate', [('x', float), ('y', float)])
+
+
+def _get_values_from_xml_element(element: etree.Element) -> List[str]:
+    """Parse an xml tag with space-separated variables."""
+    text = []  # type: List[str]
+    for e in element.itertext():
+        if e.strip():
+            text += re.split('\s+', re.sub('\n', '', e.strip()))
+    return text
+
+
+def get_calibration(file_name: Path) -> Dict[str, Any]:
+    """
+    Parse a calibration xml generated by the camera calibration tool.
+
+    (see https://docs.opencv.org/3.4.0/d7/d21/tutorial_interactive_calibration.html)
+    :param file_name: name of xml file to parse
+    :return: calibrations from the file in a dictionary
+    """
+    calibrations = {}
+    with file_name.open() as file:
+        tree = etree.parse(file)
+        root = tree.getroot()
+        for element in root:
+            if 'type_id' in element.attrib and element.attrib[
+               'type_id'] == 'opencv-matrix':
+                rows, cols = int(element.find('rows').text), int(
+                    element.find('cols').text)
+                data_type = element.find('dt').text
+                values = _get_values_from_xml_element(element.find('data'))
+                if data_type == 'd':  # doubles
+                    data = np.reshape([float(v) for v in values],
+                                      (rows, cols)).tolist()
+                else:
+                    raise ValueError('Invalid data type in xml')
+            # Integer tag names
+            elif element.tag in ['framesCount', 'cameraResolution']:
+                values = _get_values_from_xml_element(element)
+                data = [int(v) for v in values]
+            elif element.tag in ['avg_reprojection_error']:
+                values = _get_values_from_xml_element(element)
+                data = [float(v) for v in values]
+            else:
+                data = _get_values_from_xml_element(element)
+            calibrations[element.tag] = data
+    return calibrations
+
+
+@functools.lru_cache()
+def load_camera_calibrations(file_name: Path) -> Tuple[List[List[float]],
```

Comment:
Thinking again about the types this returns, I wonder whether there's a better way to use the type system to help with the return types here. As it stands we get some benefit, but it's limited.

One thing I'd wondered about doing (for the homography matrix) in the previous code was to introduce custom types for the tricky types we have from external sources (such as numpy). I think these would need explicit casts into them, though I think that that's ok as they're opaque to most of the things which are just passing them around anyway.

Certainly, it could help here (and in `calculate_transforms` below) since we're not just after any list of lists of floats, we're actually after a specific shape of them. As such we could declare our own `CameraMatrix` and `DistanceCoefficients` types and then returning them here and then require those in `calculate_transforms`.

This is definitely something we can leave for now, though I wanted to record this somewhere.